[DOC] Add new section on Kafka Bridge

laidan6000 opened the pull request:
### Type of change
- Documentation

### Description

This pull request adds a new chapter on the Strimzi Kafka Bridge to introduce the Bridge and provide information on clients, data formats, encryption, etc.

The content links to the new content in #1665 where appropriate.

Deploying the Bridge will be documented in a separate PR as this will be documented in chapter 2, Getting Started.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

seanhuck commented on documentation/book/assembly-using-the-kafka-bridge.adoc:
```diff
@@ -0,0 +1,27 @@
+[id='assembly-using-the-kafka-bridge-{context}']
+= Using the {ProductName} Kafka Bridge
+
+This chapter provides an overview of the {ProductName} Kafka Bridge and helps you get started with using the API. To view the API documentation including example requests and responses, see LINK **{ProductName} Kafka Bridge API resources**. For information on how to configure the Kafka Bridge after deployment, see LINK **{ProductName} Kafka Bridge Configuration**.
```
Remove "with" in the first sentence.

seanhuck commented on documentation/book/con-overview-of-kafka-bridge.adoc:
```diff
@@ -0,0 +1,58 @@
+
+The {ProductName} Kafka Bridge provides a REST API for integrating HTTP-based clients with a Kafka cluster running on {ProductPlatformName}. You can use it to integrate clients that do not have a Kafka client library and so cannot transfer data using the native Kafka protocol.
+
+The Kafka Bridge API consists of two resources -- `consumers` and `topics` -- that together provide several endpoints for interacting with Kafka consumers and producers. You can use the API to: 
```
@laidan6000 
```suggestion
The Kafka Bridge API consists of two resources -- `consumers` and `topics` -- that together provide several endpoints to interact with Kafka consumers and producers. You can use the API to: 
```

ppatierno commented on documentation/book/assembly-using-the-kafka-bridge.adoc:
```diff
@@ -0,0 +1,27 @@
+
+ifdef::TechPreview[]
+IMPORTANT: The {ProductName} Kafka Bridge is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service-level agreements (SLAs) and might not be functionally complete; therefore, Red Hat does not recommend implementing any Technology Preview features in production environments. This Technology Preview feature provides early access to upcoming  ...
+endif::TechPreview[]
```
I would mention that with TP the API could change in the future and even breaking changes could be possible.

laidan6000 replied:
OK, but this is standard boilerplate text. I would need to find out if we are allowed to change it.

laidan6000 replied:
I guess the note about breaking changes also applies to the upstream docs?
I've removed the TP notice from the upstream.

ppatierno commented on documentation/book/assembly-using-the-kafka-bridge.adoc:
```diff
@@ -0,0 +1,27 @@
+[id='assembly-using-the-kafka-bridge-{context}']
+= Using the {ProductName} Kafka Bridge
+
+This chapter provides an overview of the {ProductName} Kafka Bridge and helps you get started with using the API. To view the API documentation including example requests and responses, see LINK **{ProductName} Kafka Bridge API resources**. For information on how to configure the Kafka Bridge after deployment, see LINK **{ProductName} Kafka Bridge Configuration**.
```
I would say REST API in the first sentence at least, not repeating every time API is mentioned.

laidan6000 replied:
Good point.

ppatierno commented on documentation/book/con-overview-of-kafka-bridge.adoc:
```diff
@@ -0,0 +1,58 @@
+
+The Kafka Bridge API consists of two resources -- `consumers` and `topics` -- that together provide several endpoints for interacting with Kafka consumers and producers. You can use the API to: 
+
+* Send messages to a topic or partition.
```
When you send message to a topic, you are always sending message to a partition in that topic so the "or" should not be there.

laidan6000 replied:
That's why my brain seems to prefer the term "topic partition"! :D

I'll change to "topic" in all cases like this.

ppatierno commented on documentation/book/con-overview-of-kafka-bridge.adoc:
```diff
@@ -0,0 +1,58 @@
+
+* Commit a list of consumer offsets.
+
+* Perform seek operations on consumers, so that they start receiving messages from the first or last offset position, or a given offset position.
```
Actually the seek operations are executed "on a topic partition" by consumers but not "on consumers".

laidan6000 replied:
Changed to: "Seek on a partition, so that the consumer starts receiving messages from the first or last offset position, or a given offset position."

scholzj commented on documentation/book/assembly-using-the-kafka-bridge.adoc:
```diff
@@ -0,0 +1,27 @@
+This chapter provides an overview of the {ProductName} Kafka Bridge and helps you get started with using the API. To view the API documentation including example requests and responses, see LINK **{ProductName} Kafka Bridge API resources**. For information on how to configure the Kafka Bridge after deployment, see LINK **{ProductName} Kafka Bridge Configuration**.
+
+ifdef::TechPreview[]
+IMPORTANT: The {ProductName} Kafka Bridge is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service-level agreements (SLAs) and might not be functionally complete; therefore, Red Hat does not recommend implementing any Technology Preview features in production environments. This Technology Preview feature provides early access to upcoming  ...
```
This should not be in the upstream documentation.

laidan6000 replied:
OK. It will be handled via a downstream assembly.

laidan6000 commented:
I decided to create a new concept on "Requests to the Kafka Bridge". Please review the new content on "Data formats and headers": documentation/book/con-requests-kafka-bridge.adoc

ppatierno commented on documentation/book/con-requests-kafka-bridge.adoc:
```diff
@@ -0,0 +1,100 @@
+Content-Type: application/vnd.kafka.v2+json
+----
+
+* When performing producer operations, `POST` requests must provide the following `Content-Type` header specifying the *embedded data format* of the consumer, either `json` or `binary`:
```
I would not say "of the consumer". The producer has the freedom to send the data in the embedded format which makes sense to the application. It's up to the consumer to adhere to such format in order to get the data and understanding its content.

ppatierno commented on documentation/book/con-requests-kafka-bridge.adoc:
```diff
@@ -0,0 +1,100 @@
+[source,http,subs=+quotes]
+----
+Content-Type: application/vnd.kafka._embedded-data-format_.v2+json
+----
```
maybe adding two examples with json and binary embedded data format?

ppatierno commented on documentation/book/con-requests-kafka-bridge.adoc:
```diff
@@ -0,0 +1,100 @@
+
+=== Accept headers
+
+After creating a consumer, all subsequent requests must provide an `Accept` header in the following format:
```
it involves subsequent requests for getting messages but not for subscribing, deleting the consumer, etc...

ppatierno commented on documentation/book/con-requests-kafka-bridge.adoc:
```diff
@@ -0,0 +1,100 @@
+Where the `embedded-data-format` is the embedded data format of the consumer: either `json` or `binary`. The `+json` query parameter is not changed. 
+
+For example, when committing a list of consumer offsets for a consumer that has an embedded data format of JSON, include this Accept header:
+
```
committing an offset is a POST operation so you have to use Content-Type not Accept.

laidan6000 commented on documentation/book/assembly-using-the-kafka-bridge.adoc:
```diff
@@ -0,0 +1,27 @@
+
+This chapter provides an overview of the {ProductName} Kafka Bridge and helps you get started using the REST API.
+
+* To view the API documentation including example requests and responses, see LINK **{ProductName} Kafka Bridge API resources**.
```
I'll add this wording with the link to the API reference docs on strimzi.io.

laidan6000 commented on documentation/book/con-requests-kafka-bridge.adoc:
```diff
@@ -0,0 +1,100 @@
+application/vnd.kafka.[embedded_format].[api_version]+[serialization_format]
+
+
+== HTTP Verbs
```
I'll remove.

laidan6000 commented on documentation/book/con-overview-of-kafka-bridge.adoc:
```diff
@@ -0,0 +1,42 @@
+
+After the Kafka Bridge is deployed, the Cluster Operator creates a Deployment, Service, and Pod in your {ProductPlatformName} cluster, each named `strimzi-kafka-bridge`.
+
+== Supported clients
```
Agreed. I've create a new concept module for this.

laidan6000 commented:
- Removed the unfinished modules and added cross-refs to #1668 where appropriate, e.g. the deployment instructions. I decided to remove the "Sending Requests" module from the scope of this PR.
- Incorporated Jakub's content on "Securing the HTTP Kafka Bridge".
- Added a new module for Jakub's content on "Accessing the HTTP Kafka Bridge from outside of Kubernetes/Openshift"

@ppatierno - Please can you give this a final check and confirm if it's ready to merge? See http://file.rdu.redhat.com/~dlaing/master.html#assembly-using-the-kafka-bridge-str

laidan6000 commented:
@scholzj - Please can you give this a second review and confirm if it's ready to merge? I made the change in terminology to "HTTP clients" as you suggested in PR#51. 

I've included your content on " Securing the HTTP Kafka Bridge" and "Accessing the HTTP Kafka Bridge from outside of Kubernetes/Openshift". Thanks.

scholzj commented on documentation/book/con-accessing-kafka-bridge-from-outside.adoc:
```diff
@@ -0,0 +1,28 @@
+----
+  # ...
+  selector:
+    strimzi.io/cluster: kafka-bridge-name
```
The `kafka-bridge-name` has to be replaced by the actual name of the CR. Do we will mark the things which should be replaced as `<kafka-bridge-name>`? Or just mark it with this `<1>` and explain below it how it should be replaced like you do on other places.

laidan6000 replied:
I think we're a bit inconsistent in our approach. I like using annotations in the YAML examples.  I'll add a `<1>` annotation and an explanation.

scholzj commented on documentation/book/con-supported-clients-kafka-bridge.adoc:
```diff
@@ -0,0 +1,17 @@
+
+Internal clients can access the Kafka Bridge on the host and port defined in the `KafkaBridge` custom resource. External clients can access the Kafka Bridge by using an {OpenShiftName} route or a LoadBalancer Service.
+
+For more information on configuring the host and port for the `KafkaBridge` resource, see xref:ref-kafka-bridge-http-configuration-{context}[].
```
Should we link here also to `con-accessing-kafka-bridge-from-outside-{context}`?

laidan6000 replied:
Agreed. I'll add an Additional resources section to the module.
